src/macaron/repo_finder/provenance_extractor.py

```diff
@@ -0,0 +1,274 @@
+
+    if not commit:
+        raise ProvenanceExtractionException("Failed to extract commit hash from provenance.")
+
+    return repo, commit
+
+
+def _extract_from_witness_provenance(payload: InTotoV01Payload) -> tuple[str, str]:
+    """Extract the repository and commit metadata from the witness provenance file found at the passed path.
+
+    To successfully return the commit and repository URL, the payload must respectively contain a Git attestation, and
+    either a GitHub or GitLab attestation.
+
+    Parameters
+    ----------
+    payload: InTotoPayload
+        The payload to extract from.
+
+    Returns
+    -------
+    tuple[str, str]
+        The repository URL and commit hash if found, a pair of empty strings otherwise.
+    """
+    predicate: dict[str, JsonType] | None = payload.statement.get("predicate")
+    if not predicate:
+        raise ProvenanceExtractionException("No predicate in payload statement.")
+
+    attestations = json_extract(predicate, ["attestations"], list)
+    commit = ""
+    repo = ""
+    for entry in attestations:
+        if not isinstance(entry, dict):
+            continue
+        entry_type = entry.get("type")
+        if not entry_type:
+            continue
+        if entry_type.startswith("https://witness.dev/attestations/git/"):
+            commit = json_extract(entry, ["attestation", "commithash"], str)
+        elif entry_type.startswith("https://witness.dev/attestations/gitlab/") or entry_type.startswith(
+            "https://witness.dev/attestations/github/"
+        ):
+            repo = json_extract(entry, ["attestation", "projecturl"], str)
+
+    if not commit or not repo:
+        raise ProvenanceExtractionException("Could not extract repo and commit from provenance.")
+
+    return repo, commit
+
+
+def _extract_commit_from_digest_set(digest_set: dict[str, JsonType], valid_algorithms: list[str]) -> str:
+    """Extract the commit from the passed DigestSet.
+
+    The DigestSet is an in-toto object that maps algorithm types to commit hashes (digests).
+    """
+    if len(digest_set.keys()) > 1:
+        logger.debug("DigestSet contains multiple algorithms: %s", digest_set.keys())
+
+    for key in digest_set:
+        if key in valid_algorithms:
+            value = digest_set.get(key)
+            if isinstance(value, str):
+                return value
+    raise ProvenanceExtractionException(f"No valid digest in digest set: {digest_set.keys()} not in {valid_algorithms}")
+
+
+def _clean_spdx(uri: str) -> str:
+    """Clean the passed SPDX URI and return the normalised URL it represents.
+
+    A SPDX URI has the form: git+https://example.com@refs/heads/main
+    """
+    url, _, _ = uri.lstrip("git+").rpartition("@")
+    return url
+
+
+class JsonExtractionException(MacaronError):
+    """When there is an error while extracting from JSON."""
+
+
+T = TypeVar("T", bound=JsonType)
+
+
+def json_extract(entry: JsonType, keys: list[str], type_: type[T]) -> T:
```

Comment:
This is a useful function that can be used outside this module.

Can you please move it to `src/macaron/json_tools.py` module?